sphinx/io.py

```diff
@@ -153,30 +152,45 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
 
 
-def read_doc(app: "Sphinx", env: BuildEnvironment, filename: str) -> nodes.document:
+def read_doc(publisher: Publisher, docname: str, filename: str) -> nodes.document:
```

Comment:
I'm not sure about the responsibility of the new `read_doc()`. This only calls the `publish()` method of the given publisher. I feel there is no reason to be separated from the env class. Is it better to merge this into it?